beluga/test/beluga/algorithm/test_unscented_transform.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <cmath>
+
+#include <Eigen/Core>
+
+#include "beluga/algorithm/unscented_transform.hpp"
+
+struct UnscentedTransformTests : public ::testing::Test {};
+
+TEST_F(UnscentedTransformTests, Identity) {
+  const double k_tolerance = 1e-4;
+  const Eigen::Vector2d mean{12.3, 7.6};
+  Eigen::Matrix<double, 2, 2> cov;
+  cov << 1.44, 0.,  //
+      0., 2.89;
+
+  const auto [transformed_mean, transformed_cov] =
+      beluga::unscented_transform(mean, cov, [](const Eigen::Vector2d& in) -> Eigen::Vector2d { return in; });
+
+  ASSERT_TRUE(transformed_mean.isApprox(mean, k_tolerance)) << mean.transpose();
+  ASSERT_TRUE(transformed_cov.isApprox(cov, k_tolerance)) << cov.transpose();
+}
+
+TEST_F(UnscentedTransformTests, Scaling) {
+  const double k_tolerance = 1e-4;
+  const Eigen::Vector2d mean{12.3, 7.6};
+
+  const double k_stdev_x = 1.2;
+  const double k_stdev_y = 1.7;
+  Eigen::Matrix<double, 2, 2> cov;
+  cov << k_stdev_x * k_stdev_x, 0.,  //
+      0., k_stdev_y * k_stdev_y;
+
+  const auto [transformed_mean, transformed_cov] =
+      beluga::unscented_transform(mean, cov, [](const Eigen::Vector2d& in) -> Eigen::Vector2d { return 2 * in; });
+
+  Eigen::Matrix<double, 2, 2> expected_cov;
+  expected_cov << (2. * k_stdev_x) * (2. * k_stdev_x), 0.,  //
+      0., (2. * k_stdev_y) * (2. * k_stdev_y);
+
+  ASSERT_TRUE(transformed_mean.isApprox(2. * mean, k_tolerance)) << mean.transpose();
+  ASSERT_TRUE(transformed_cov.isApprox(expected_cov, k_tolerance)) << transformed_cov << " \n" << expected_cov;
+}
+
+TEST_F(UnscentedTransformTests, CartesianToPolar) {
+  // This example is taken from wikipedia, see https://en.wikipedia.org/wiki/Unscented_transform .
+  const double k_tolerance = 1e-2;
+  const Eigen::Vector2d mean{12.3, 7.6};
+
+  Eigen::Matrix<double, 2, 2> cov;
+  cov << 1.44, 0.,  //
+      0., 2.89;
+
+  const auto [transformed_mean, transformed_cov] =
+      beluga::unscented_transform(mean, cov, [](const Eigen::Vector2d& in) -> Eigen::Vector2d {
+        return Eigen::Vector2d{std::hypot(in.x(), in.y()), std::atan2(in.y(), in.x())};
+      });
+
+  Eigen::Matrix<double, 2, 2> expected_cov;
+  expected_cov << 1.823, 0.043,  //
+      0.043, 0.012;
+
+  const Eigen::Vector2d k_expected_mean{14.454, 0.55};
```

Comment:
@serraramiro1 qq: are the analytical forms of these too complicated to be reproduced here? I'd prefer those over magic numbers, if possible.

Reply:
This is the expected mean for the UT, the analytical form would be the UT implementation itself.
I've gotten the results directly from wikipedia page.